### PR TITLE
Change feature_index from fake Array to darray 

### PR DESCRIPTION
**Pull request by peterzhu2118**
Using a fake (malloc) RArray is not friendly for the garbage
collector. This commit changes feature_index from a RArray to a
darray.

### CODE REVIEW

File: darray.h [hunk 1]
```diff
@@ -21,19 +21,19 @@
 
 // Copy an element out of the array. Warning: not bounds checked.
 //
-// T rb_darray_get(rb_darray(T) ary, int32_t idx);
+// T rb_darray_get(rb_darray(T) ary, long idx);
```
Comment (maximecb) [inline darray.h:24]: Any reason why you're using long instead of `size_t` ?

Comment (peterzhu2118) [reply darray.h:24]: That's a good point. Ruby's Array type uses `long` for it's capacity and sizes. I'm also not 100% sure about the overflow checking implications that @XrXr has implemented when changing from a signed integer to an unsigned integer.

File: darray.h [hunk 2]
```diff
@@ -80,23 +89,33 @@
...
 // Note that NULL is a valid empty dynamic array.
 //
-// bool rb_darray_make(rb_darray(T) *ptr_to_ary, int32_t size);
+// bool rb_darray_make(rb_darray(T) *ptr_to_ary, long size);
 //
-#define rb_darray_make(ptr_to_ary, size) rb_darray_make_impl((ptr_to_ary), size, sizeof(**(ptr_to_ary)), sizeof((*(ptr_to_ary))->data[0]))
+#define rb_darray_make(ptr_to_ary, size) \
+    rb_darray_make_impl((ptr_to_ary), size, sizeof(**(ptr_to_ary)), sizeof((*(ptr_to_ary))->data[0]), calloc)
+
+// Like rb_darray_make but uses ruby_xmalloc (meaning it could be a possible GC point).
```
[the remark this answers is not in the document]
Comment (peterzhu2118) [reply darray.h]: Fixed! Thanks for catching the typo.

File: darray.h (hunk 1 above)
Comment (peterzhu2118) [reply darray.h]: After discussions with @XrXr, we've decided to use `size_t` and use Ruby's builtin `malloc` with overflow checking.

**Comment by peterzhu2118** [comment]
YJIT is currently incompatible (crashes) with malloc functions that run GC, so I've gone back to adding `_with_gc` functions that use the malloc functions provided by the GC. Once YJIT migrates to Rust, we can refactor this code and replace the default functions with the `_with_gc` functions.

**Pull request merged**